applications/MeshingApplication/CMakeLists.txt

```diff
@@ -66,15 +66,81 @@ if(${INCLUDE_MMG} MATCHES ON)
 
+if(${INCLUDE_PMMG} MATCHES ON)
+    IF(${USE_MPI} MATCHES OFF )
+        message( FATAL_ERROR "Please compile Kratos in MPI to use ParMmg by setting the flag -DUSE_MPI=ON")
+    ENDIF(${USE_MPI} MATCHES OFF )
+
+    file( GLOB KRATOS_MESHING_APPLICATION_CORE ${KRATOS_MESHING_APPLICATION_CORE}
+        ${CMAKE_CURRENT_SOURCE_DIR}/custom_utilities/parmmg/*.cpp
+        ${CMAKE_CURRENT_SOURCE_DIR}/custom_processes/parmmg/*.cpp
+    )
+
+    IF(NOT DEFINED INCLUDE_MMG)
+        message( FATAL_ERROR "MMG not included. Please add -DINCLUDE_MMG=ON to the configure file in order to compile with ParMMG")
+    ELSE(NOT DEFINED INCLUDE_MMG)
+        if(${INCLUDE_MMG} MATCHES OFF)
+            message( FATAL_ERROR " MMG disabled. Please set -DINCLUDE_MMG=ON in the configure file in order to compile with ParMMG")
+        endif(${INCLUDE_MMG} MATCHES OFF)
+    ENDIF(NOT DEFINED INCLUDE_MMG)
+
+    IF(NOT DEFINED MMG3D_LIBRARIES)
+        message( FATAL_ERROR "MMG3D not found. Please make sure MMG3D libraries are added setting the MMG build path as MMG_ROOT")
+    ENDIF(NOT DEFINED MMG3D_LIBRARIES)
+
+    add_definitions( -DINCLUDE_PMMG )
+    ## find Metis for ParMmg
+    if(NOT DEFINED USE_METIS_5)
+        message("USE_METIS_5 not defined. Setting to ON")
+        SET (USE_METIS_5 ON)
+    endif(NOT DEFINED USE_METIS_5)
+
+    IF(NOT DEFINED PMMG_LIBRARY_PREFIX)
+        SET(PMMG_LIBRARY_PREFIX "lib")
+    ENDIF(NOT DEFINED PMMG_LIBRARY_PREFIX)
+
+    IF (PMMG_ROOT)
+        IF(NOT DEFINED PMMG_LIBRARY_EXTENSION)
+            IF(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
+                IF(EXISTS "${PMMG_ROOT}/lib/${PMMG_LIBRARY_PREFIX}parmmg.dll") # Dynamic library
+                    SET(PMMG_LIBRARY_EXTENSION "dll")
+                ELSE(EXISTS "${PMMG_ROOT}/lib/${PMMG_LIBRARY_PREFIX}parmmg.dll") # Static library
+                    SET(PMMG_LIBRARY_EXTENSION "lib")
+                ENDIF(EXISTS "${PMMG_ROOT}/lib/${PMMG_LIBRARY_PREFIX}parmmg.dll")
+            ELSE(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
+                IF(EXISTS "${PMMG_ROOT}/lib/${PMMG_LIBRARY_PREFIX}parmmg.so") # Dynamic library
+                    SET(PMMG_LIBRARY_EXTENSION "so")
+                ELSE(EXISTS "${PMMG_ROOT}/lib/${PMMG_LIBRARY_PREFIX}parmmg.so") # Static library
+                    SET(PMMG_LIBRARY_EXTENSION "a")
+                ENDIF(EXISTS "${PMMG_ROOT}/lib/${PMMG_LIBRARY_PREFIX}parmmg.so")
+            ENDIF(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
+        ENDIF(NOT DEFINED PMMG_LIBRARY_EXTENSION)
+
+        IF(NOT DEFINED PMMG_INCLUDE_DIR)
+            SET(PMMG_INCLUDE_DIR ${PMMG_ROOT}/include/)
+        ENDIF(NOT DEFINED PMMG_INCLUDE_DIR)
+        IF(NOT DEFINED PMMG_LIBRARY)
+            SET(PMMG_LIBRARY ${PMMG_ROOT}/lib/${PMMG_LIBRARY_PREFIX}parmmg.${PMMG_LIBRARY_EXTENSION})
+        ENDIF(NOT DEFINED PMMG_LIBRARY)
+    ENDIF (PMMG_ROOT)
+
+    set(KRATOS_MESHING_APPLICATION_EXTRA_LIBS ${KRATOS_MESHING_APPLICATION_EXTRA_LIBS} KratosMPICore ${PMMG_LIBRARY})
```

Comment:
are you sure you cannot make it work without introducing the MPICore dependency on c++ level?
=> then you also need to import the mpi-core in the init of the application like the TrilinosApp does (in your case only in distributed runs)
otherwise you could get undefined symbols

Reply:
Compiling without KratosMPICore gives the following warnings, which I guess are coming from lib calls to parmmg contained in the utils/process?



```
pmmg_utilities.cpp:(.text._ZNK3MPI8Cartcomm5CloneEv[_ZNK3MPI8Cartcomm5CloneEv]+0x20): undefined reference to `MPI_Comm_dup'
pmmg_utilities.cpp:(.text._ZNK3MPI8Cartcomm5CloneEv[_ZNK3MPI8Cartcomm5CloneEv]+0x35): undefined reference to `MPI::Comm::Comm()'
pmmg_utilities.cpp:(.text._ZNK3MPI8Cartcomm5CloneEv[_ZNK3MPI8Cartcomm5CloneEv]+0x55): undefined reference to `MPI_Initialized'
pmmg_utilities.cpp:(.text._ZNK3MPI8Cartcomm5CloneEv[_ZNK3MPI8Cartcomm5CloneEv]+0x69): undefined reference to `ompi_mpi_comm_null'
pmmg_utilities.cpp:(.text._ZNK3MPI8Cartcomm5CloneEv[_ZNK3MPI8Cartcomm5CloneEv]+0x78): undefined reference to `MPI_Topo_test'
CMakeFiles/KratosMeshingCore.dir/custom_utilities/parmmg/pmmg_utilities.cpp.o: In function `MPI::Intracomm::Create_cart(int, int const*, bool const*, bool) const':
pmmg_utilities.cpp:(.text._ZNK3MPI9Intracomm11Create_cartEiPKiPKbb[_ZNK3MPI9Intracomm11Create_cartEiPKiPKbb]+0x171): undefined reference to `MPI_Cart_create'
pmmg_utilities.cpp:(.text._ZNK3MPI9Intracomm11Create_cartEiPKiPKbb[_ZNK3MPI9Intracomm11Create_cartEiPKiPKbb]+0x181): undefined reference to `MPI::Comm::Comm()'
pmmg_utilities.cpp:(.text._ZNK3MPI9Intracomm11Create_cartEiPKiPKbb[_ZNK3MPI9Intracomm11Create_cartEiPKiPKbb]+0x1a2): undefined reference to `MPI_Initialized'
pmmg_utilities.cpp:(.text._ZNK3MPI9Intracomm11Create_cartEiPKiPKbb[_ZNK3MPI9Intracomm11Create_cartEiPKiPKbb]+0x1b8): undefined reference to `ompi_mpi_comm_null'
pmmg_utilities.cpp:(.text._ZNK3MPI9Intracomm11Create_cartEiPKiPKbb[_ZNK3MPI9Intracomm11Create_cartEiPKiPKbb]+0x1c7): undefined reference to `MPI_Topo_test'
CMakeFiles/KratosMeshingCore.dir/custom_utilities/parmmg/pmmg_utilities.cpp.o: In function `MPI::Intercomm::Merge(bool) const':
pmmg_utilities.cpp:(.text._ZNK3MPI9Intercomm5MergeEb[_ZNK3MPI9Intercomm5MergeEb]+0x2c): undefined reference to `MPI_Intercomm_merge'
pmmg_utilities.cpp:(.text._ZNK3MPI9Intercomm5MergeEb[_ZNK3MPI9Intercomm5MergeEb]+0x39): undefined reference to `MPI::Comm::Comm()'
pmmg_utilities.cpp:(.text._ZNK3MPI9Intercomm5MergeEb[_ZNK3MPI9Intercomm5MergeEb]+0x59): undefined reference to `MPI_Initialized'
pmmg_utilities.cpp:(.text._ZNK3MPI9Intercomm5MergeEb[_ZNK3MPI9Intercomm5MergeEb]+0x68): undefined reference to `ompi_mpi_comm_null'
pmmg_utilities.cpp:(.text._ZNK3MPI9Intercomm5MergeEb[_ZNK3MPI9Intercomm5MergeEb]+0xa1): undefined reference to `MPI_Comm_test_inter'
CMakeFiles/KratosMeshingCore.dir/custom_utilities/parmmg/pmmg_utilities.cpp.o: In function `MPI::Intracomm::Create(MPI::Group const&) const':
pmmg_utilities.cpp:(.text._ZNK3MPI9Intracomm6CreateERKNS_5GroupE[_ZNK3MPI9Intracomm6CreateERKNS_5GroupE]+0x2f): undefined reference to `MPI_Comm_create'
pmmg_utilities.cpp:(.text._ZNK3MPI9Intracomm6CreateERKNS_5GroupE[_ZNK3MPI9Intracomm6CreateERKNS_5GroupE]+0x3c): undefined reference to `MPI::Comm::Comm()'
pmmg_utilities.cpp:(.text._ZNK3MPI9Intracomm6CreateERKNS_5GroupE[_ZNK3MPI9Intracomm6CreateERKNS_5GroupE]+0x5c): undefined reference to `MPI_Initialized'
pmmg_utilities.cpp:(.text._ZNK3MPI9Intracomm6CreateERKNS_5GroupE[_ZNK3MPI9Intracomm6CreateERKNS_5GroupE]+0x6b): undefined reference to `ompi_mpi_comm_null'
pmmg_utilities.cpp:(.text._ZNK3MPI9Intracomm6CreateERKNS_5GroupE[_ZNK3MPI9Intracomm6CreateERKNS_5GroupE]+0xa1): undefined reference to `MPI_Comm_test_inter'
CMakeFiles/KratosMeshingCore.dir/custom_utilities/parmmg/pmmg_utilities.cpp.o: In function `MPI::Intracomm::Split(int, int) const':
pmmg_utilities.cpp:(.text._ZNK3MPI9Intracomm5SplitEii[_ZNK3MPI9Intracomm5SplitEii]+0x2b): undefined reference to `MPI_Comm_split'
pmmg_utilities.cpp:(.text._ZNK3MPI9Intracomm5SplitEii[_ZNK3MPI9Intracomm5SplitEii]+0x38): undefined reference to `MPI::Comm::Comm()'
pmmg_utilities.cpp:(.text._ZNK3MPI9Intracomm5SplitEii[_ZNK3MPI9Intracomm5SplitEii]+0x58): undefined reference to `MPI_Initialized'
pmmg_utilities.cpp:(.text._ZNK3MPI9Intracomm5SplitEii[_ZNK3MPI9Intracomm5SplitEii]+0x67): undefined reference to `ompi_mpi_comm_null'
pmmg_utilities.cpp:(.text._ZNK3MPI9Intracomm5SplitEii[_ZNK3MPI9Intracomm5SplitEii]+0x99): undefined reference to `MPI_Comm_test_inter'
CMakeFiles/KratosMeshingCore.dir/custom_utilities/parmmg/pmmg_utilities.cpp.o:(.data.rel.ro._ZTVN3MPI8DatatypeE[_ZTVN3MPI8DatatypeE]+0x78): undefined reference to `MPI::Datatype::Free()'
CMakeFiles/KratosMeshingCore.dir/custom_utilities/parmmg/pmmg_utilities.cpp.o:(.data.rel.ro._ZTVN3MPI3WinE[_ZTVN3MPI3WinE]+0x48): undefined reference to `MPI::Win::Free()'


```

Reply:
ok then it seems like ParMMG has MPI calls in the headers that you include. @roigcarlo IMO this is bad design on the ParMMG side, I would only have the MPI-calls in the cpp so that the linking to MPI is done properly when compiling the library itself.

In any case this would probably require modifications on their side, then I guess you still need to link against the MPICore (which links against the MPI-libs which makes the warnings disappear) for the moment, maybe after the release this can be improved

Still on the KratosSide I would carefully consider if you need to explicitly use code from the MPICore for the above mentioned issues

Reply:
BTW just to add that I am also not an expert in this linking and dependency management stuff, hence my opinion might be wrong :D
@roigcarlo please correct me if what I wrote doesn't make sense ;)

Reply:
Sorry for the delay. As @philbucher says, this is a implementation detail on mmg side. There is a alternative that we could follow like the on we do with the Trilinos extensions for the structural and the fluid. 

For my side, that would be good to have, but its no reason to block this right now. So go ahead with the merge.

Reply:
Ok perfect